[GNIP #3944] Make GeoFence efficient for GeoNode instances with a large number of layers

afabiani opened the pull request:
This PR also would like to address the issue https://github.com/GeoNode/geonode/issues/4125
"GeoFence cache seems to be refreshed every time a layer is uploaded"

capooti commented on geonode/geoserver/management/commands/sync_geonode_layers.py:
```diff
@@ -43,10 +43,10 @@ def sync_geonode_layers(ignore_errors, filter, username):
             layer.set_permissions(perm_spec)
             # recalculate the layer statistics
             set_attributes_from_geoserver(layer, overwrite=True)
-            layer.save()
```
This command is not only related to permissions, but to general synchronization between GeoServer/GeoFence/Django. I believe without this layer.save() thumbnails and styles won't be synced

afabiani replied:
@capooti this was my intention. I mean, it seems to me dangerous and also a waste of time to regenerate all the thumbnails for the layers synced. It seems to me that the thumbnail, except the first time, should be manually updated by the user from the gui or at most we should envisage a specific management command for this.

capooti replied:
well this command was thought as a sort of updatelayers alternative version which doesn't use GetCapabilities (which can be broken and it is broken and really slow in our case). The use case for us was to use it as a replacement to it.
We are also using it to migrate GeoNode itself, which explains why we want to update the thumbnails - in fact they are being generated for the first time in such a case. How about to add it as a command option that launch layer.save() if enabled?

afabiani replied:
@capooti yes, good point. Make sense to add an explicit option which will warn users to eventually provide a filter whenever they want to explicitly update the thumbnails too. I will push the change in a few unless there are other objections.

giohappy replied:
@afabiani I would add the following optional arguments: --updatepermissions, --updatethumbnails
If none are set all actions will be executed, otherwise only the ones set are executed.

capooti replied:
great! thanks @afabiani and @giohappy

t-book commented:
I´ve just tested with a large number of layers (12.500). It´s  working much smoother now (https://github.com/GeoNode/geonode/issues/4125) Thanks @afabiani 

+1 for @giohappy suggestions regarding --updatepermissions, --updatethumbnails

capooti commented:
@t-book that is great to know. How long takes a layer upload for you?

t-book commented:
> @t-book that is great to know. How long takes a layer upload for you?

@capooti  ~10 – 20 seconds but this is not a valid measurement as 
- I´ve tested locally 
- Tested with very small test-shape-files (3-5 points each)

So basically I can say for sure:

**Old behavior:**
- changing permissions for one layer affected rules for all layers (**slow not usable for production**)

**New behavior:**
- changing permissions for one  layer only affects specific rules

I will repeat/do more testing on a remote staging instance with around 4000 layers. (Unfortunately at earliest next week)

giohappy commented on geonode/geoserver/management/commands/sync_geonode_layers.py:
```diff
@@ -41,12 +41,17 @@ def sync_geonode_layers(ignore_errors, filter, username):
-            layer.set_permissions(perm_spec)
-            # recalculate the layer statistics
-            set_attributes_from_geoserver(layer, overwrite=True)
+            if updatepermissions or not updatethumbnails:
```
@afabiani I would make these arguments True by default and simply check is they're set, without any permutive check. I.e. if none are explicitly set by the user, ot both are set, both the actions should be performed.

afabiani replied:
@giohappy done

giohappy replied:
@afabiani what was something between this and the previous. Ok the defaults to true but if any of them is set the other must be False, unless they are explicitely true.

afabiani commented:
@capooti @t-book with the latest commits I have updated the management command as suggested by @giohappy and added a few more improvements which should allow GeoNode to work with a very large number of datasets.

Currently, on a test instance with 22K layers the behavior is good.
The only things I noticed is that the api for the categories often seems to fail due to a memory exception. Not sure if it is a DB issue or, most probably, a memory leak with tastypie. This will need more investigation and it is bread for another issue.

I kindly suggest, in a production environment, to configure uWSGI wisely.

Typically try to have many processes with 1/2 threads each, since threads consume a lot of memory.

Also, be careful about the dynamic memory allocation. Spawning a lot of processes may easily consume all the available memory, therefore limit them. Moreover, in the case, GeoServer is running on the same machine, consider also the max heap memory needed by the Virtual Machine. If no more RAM is available, the JVM will crash.

t-book commented:
Thanks a lot @afabiani ! Good point having an eye on uWSGI memory usage. In production, I normally set threads depending on available CPUs. 

Regarding memory I would add lowering logging to critical issues only ( `disable-logging = True` ) .
Also it might be worth checking process refreshing f.e. `max-worker-lifetime = 30`. The newer the process the faster it is (viewer memory management tasks). 

Thanks again!
